politeiad/backendv2/tstorebe/plugins/ticketvote/cmds.go

```diff
@@ -0,0 +1,2976 @@
+				// No commitment address; continue
+			case *scriptPubKey.CommitAmt > bestAmt:
+				// New largest commitment address found
+				bestAddr = scriptPubKey.Addresses[0]
+				bestAmt = *scriptPubKey.CommitAmt
+			}
+		}
+		if bestAddr == "" || bestAmt == 0.0 {
+			addrErr = fmt.Errorf("no largest commitment address found")
+		}
+
+		// Store result
+		addrs[tx.TxID] = commitmentAddr{
+			addr: bestAddr,
+			err:  addrErr,
+		}
+	}
+
+	return addrs, nil
+}
+
+// startReply fetches all required data and returns a StartReply.
+func (p *ticketVotePlugin) startReply(duration uint32) (*ticketvote.StartReply, error) {
+	// Get the best block height
+	bb, err := p.bestBlock()
+	if err != nil {
+		return nil, fmt.Errorf("bestBlock: %v", err)
+	}
+
+	// Find the snapshot height. Subtract the ticket maturity from the
+	// block height to get into unforkable territory.
+	ticketMaturity := uint32(p.activeNetParams.TicketMaturity)
+	snapshotHeight := bb - ticketMaturity
+
+	// Fetch the block details for the snapshot height. We need the
+	// block hash in order to fetch the ticket pool snapshot.
+	bd := dcrdata.BlockDetails{
+		Height: snapshotHeight,
+	}
+	payload, err := json.Marshal(bd)
+	if err != nil {
+		return nil, err
+	}
+	reply, err := p.backend.PluginRead(nil, dcrdata.PluginID,
+		dcrdata.CmdBlockDetails, string(payload))
+	if err != nil {
+		return nil, fmt.Errorf("PluginRead %v %v: %v",
+			dcrdata.PluginID, dcrdata.CmdBlockDetails, err)
+	}
+	var bdr dcrdata.BlockDetailsReply
+	err = json.Unmarshal([]byte(reply), &bdr)
+	if err != nil {
+		return nil, err
+	}
+	if bdr.Block.Hash == "" {
+		return nil, fmt.Errorf("invalid block hash for height %v", snapshotHeight)
+	}
+	snapshotHash := bdr.Block.Hash
+
+	// Fetch the ticket pool snapshot
+	tp := dcrdata.TicketPool{
+		BlockHash: snapshotHash,
+	}
+	payload, err = json.Marshal(tp)
+	if err != nil {
+		return nil, err
+	}
+	reply, err = p.backend.PluginRead(nil, dcrdata.PluginID,
+		dcrdata.CmdTicketPool, string(payload))
+	if err != nil {
+		return nil, fmt.Errorf("PluginRead %v %v: %v",
+			dcrdata.PluginID, dcrdata.CmdTicketPool, err)
+	}
+	var tpr dcrdata.TicketPoolReply
+	err = json.Unmarshal([]byte(reply), &tpr)
+	if err != nil {
+		return nil, err
+	}
+	if len(tpr.Tickets) == 0 {
+		return nil, fmt.Errorf("no tickets found for block %v %v",
+			snapshotHeight, snapshotHash)
+	}
+
+	// The start block height has the ticket maturity subtracted from
+	// it to prevent forking issues. This means we the vote starts in
+	// the past. The ticket maturity needs to be added to the end block
+	// height to correct for this.
+	endBlockHeight := snapshotHeight + duration + ticketMaturity
+
+	return &ticketvote.StartReply{
+		StartBlockHeight: snapshotHeight,
+		StartBlockHash:   snapshotHash,
+		EndBlockHeight:   endBlockHeight,
+		EligibleTickets:  tpr.Tickets,
+	}, nil
+}
+
+func (p *ticketVotePlugin) cmdAuthorize(treeID int64, token []byte, payload string) (string, error) {
```

Comment:
This function needs doco. I had to read it several times to even grok what was going on.

Reply:
Done.

https://github.com/lukebp/politeia/blob/tlog/politeiad/backendv2/tstorebe/plugins/ticketvote/cmds.go#L45